src/OrleansRuntime/GrainDirectory/RemoteGrainDirectory.cs

```diff
@@ -22,313 +20,52 @@ internal RemoteGrainDirectory(LocalGrainDirectory r, GrainId id)
-                throw new OrleansException("Silo " + router.MyAddress + " is not the owner of the grain " + grain + " Owner=" + owner);
-            }
+            return router.DeleteGrainAsync(grainId, hopCount);
         }
 
-        public async Task<Tuple<List<Tuple<SiloAddress, ActivationId>>, int>> LookUp(GrainId grain, int retries)
+        public async Task<AddressesAndTag> LookupAsync(GrainId grainId, int hopCount)
         {
-            router.RemoteLookupsReceived.Increment();
-
-            // validate that this grain should be stored in our partition
-            SiloAddress owner = router.CalculateTargetSilo(grain, false);
-            if (router.MyAddress.Equals(owner))
-            {
-                router.LocalDirectoryLookups.Increment();
-                // It can happen that we cannot find the grain in our partition if there were 
-                // some recent changes in the membership. Return empty list in such case (and not null) to avoid
-                // NullReference exceptions in the code of invokers
-                Tuple<List<Tuple<SiloAddress, ActivationId>>, int> res = partition.LookUpGrain(grain);
-                if (res != null)
-                {
-                    router.LocalDirectorySuccesses.Increment();
-                    return res;
-                }
-
-                return new Tuple<List<Tuple<SiloAddress, ActivationId>>, int>(new List<Tuple<SiloAddress, ActivationId>>(), GrainInfo.NO_ETAG);
-            }
-
-            if (retries <= 0)
-                throw new OrleansException("Silo " + router.MyAddress + " is not the owner of the grain " + 
-                    grain + " Owner=" + owner);
-
-            if (logger.IsVerbose2) logger.Verbose2("Retry " + retries + " RemoteGrainDirectory.LookUp for Grain=" + grain + " at Owner=" + owner);
-            
-            PrepareForRetry(retries);
-            await Task.Delay(RETRY_DELAY);
-
-            SiloAddress o = router.CalculateTargetSilo(grain, false);
-            if (router.MyAddress.Equals(o))
-            {
-                router.LocalDirectoryLookups.Increment();
-                var res = partition.LookUpGrain(grain);
-                if (res == null)
-                    return new Tuple<List<Tuple<SiloAddress, ActivationId>>, int>(
-                        new List<Tuple<SiloAddress, ActivationId>>(), GrainInfo.NO_ETAG);
-
-                router.LocalDirectorySuccesses.Increment();
-                return res;
-            }
-            router.RemoteLookupsSent.Increment();
-            return await GetDirectoryReference(o).LookUp(grain, retries - 1);
+            return await router.LookupAsync(grainId, hopCount);
         }
 
-        public Task<List<Tuple<GrainId, int, List<Tuple<SiloAddress, ActivationId>>>>> LookUpMany(List<Tuple<GrainId, int>> grainAndETagList, int retries)
+        public async Task<List<Tuple<GrainId, int, List<ActivationAddress>>>> LookUpMany(List<Tuple<GrainId, int>> grainAndETagList)
```

Comment:
this should go back to removing the async keyword, as there are no longer any awaits in here.